bookwyrm/static/js/bookwyrm.js

```diff
@@ -0,0 +1,308 @@
+        }
+
+        // Show/hide container.
+        let container = document.getElementById('hide-' + targetId);
+
+        if (container) {
+            this.toggleContainer(container, pressed);
+        }
+
+        // Check checkbox, if appropriate.
+        let checkbox = trigger.dataset['controls-checkbox'];
+
+        if (checkbox) {
+            this.toggleCheckbox(checkbox, pressed);
+        }
+
+        // Set focus, if appropriate.
+        let focus = trigger.dataset['focus-target'];
+
+        if (focus) {
+            this.toggleFocus(focus);
+        }
+    }
+
+    /**
+     * Show or hide menus.
+     *
+     * @param  {Event} event
+     * @return {undefined}
+     */
+    toggleMenu(trigger, targetId) {
+        let expanded = trigger.getAttribute('aria-expanded') == 'false';
+
+        trigger.setAttribute('aria-expanded', expanded);
+
+        if (targetId) {
+            let target = document.getElementById(targetId);
+
+            this.addRemoveClass(target, 'is-active', expanded);
+        }
+    }
+
+    /**
+     * Show or hide generic containers.
+     *
+     * @param  {object}  container - DOM node
+     * @param  {boolean} pressed   - Is the trigger pressed?
+     * @return {undefined}
+     */
+    toggleContainer(container, pressed) {
+        this.addRemoveClass(container, 'is-hidden', pressed);
+    }
+
+    /**
+     * Check or uncheck a checbox.
+     *
+     * @param  {object}  checkbox - DOM node
+     * @param  {boolean} pressed  - Is the trigger pressed?
+     * @return {undefined}
+     */
+    toggleCheckbox(checkbox, pressed) {
+        document.getElementById(checkbox).checked = !!pressed;
+    }
+
+    /**
+     * Give the focus to an element.
+     * Only move the focus based on user interactions.
+     *
+     * @param  {string} nodeId - ID of the DOM node to focus (button, link…)
+     * @return {undefined}
+     */
+    toggleFocus(nodeId) {
+        let node = document.getElementById(nodeId);
+
+        node.focus();
+
+        setTimeout(function() {
+            node.selectionStart = node.selectionEnd = 10000;
+        }, 0);
+    }
+
+    /**
+     * Make a request and update the UI accordingly.
+     * This function is used for boosts and favourites.
+     *
+     * @param  {Event} event
+     * @return {undefined}
+     */
+    interact(event) {
+        event.preventDefault();
+
+        const bookwyrm = this;
+
+        let allTriggers = document.querySelectorAll(`.${event.target.dataset.id}`);
+
+        // Change icon to show ongoing activity on the current UI.
+        allTriggers.forEach(node => bookwyrm.addRemoveClass(
+            node,
+            'is-processing',
+            true
+        ));
+
+        this.ajaxPost(event.target)
+            .finally(() => {
+                // Change icon to remove ongoing activity on the current UI.
+                allTriggers.forEach(node => bookwyrm.addRemoveClass(
+                    node,
+                    'is-processing',
```

Comment:
I didn’t notice this was also used on the Follow button. I’ll check that out; thanks!

As mentioned in my [previous comment](https://github.com/mouse-reeve/bookwyrm/pull/883#discussion_r609381969), _Immediately_ is always very relative to the context one person is in. :)

Reply:
For submit button, they should really be `disabled` when a request is made, to avoid the behaviour you’re mentioning. This would act as both an indicator something is happening and a way to avoid multiple requests.

Reply:
Is there a specific reason for the Response of the _Follow Request_ to be the HTML document of the followed user?

Reply:
I made a [commit](https://github.com/mouse-reeve/bookwyrm/pull/883/commits/83bd417878a0aa7729f27459a0927352ab4edff1) that should fix the issues you’re mentioning. We could also remove the animation all the same as the disabled state of the _submitter_ indicates the processing.

Reply:
The follow request request response is html is that the server receives form submissions through both ajax requests and regular form actions, and handles them as form requests. If javascript was disabled, the Follow form would be submitted normally and the action would still work, just with a page reload.

Reply:
Ha, I see… Awesome. :)

The request could be proxied to serve different responses based on the parameters provided in the request. This way, we can either load the page, or just a JSON response. That would reduce the response time and the load.

Reply:
Yeah, I think flagging ajax requests and responding accordingly would be a great change, like you said, it would be faster and it would produce meaningful error states